sherpa/astro/io/__init__.py

```diff
@@ -610,10 +610,24 @@ def read_pha(arg, use_errors=False, use_background=False):
 def _pack_table(dataset):
     """Identify the columns in the data.
 
+    This relies on the _fields attribute containing the
+    data columns, and _extra_fields the extra information
+    (other than the name column).
+
     """
     names = dataset._fields
-    data = {name: getattr(dataset, name) for name in names}
-    return data, names
+    data = {}
+    for name in dataset._fields:
+        if name == 'name':
+            continue
+
+        val = getattr(dataset, name)
+        if val is  None:
+            continue
+
+        data[name] = val
+
+    return data, list(data.keys())
```

Comment:
Actually, the one-line version is too complicated here because these are not astropy table objects but Data objects and you can not access them as `dataset[n]`. So you's have to say something like

    return {n: getattr(dataset, n)
               for n in dataset._fields
               if (n != 'name') and (getattr(dataset, n) is not None)}

and I think this is too large to be inlined.

Reply:
I haven't made this change as I don't think it's worth it.